src/EventStore.Client/GossipChannelSelector.cs

```diff
@@ -0,0 +1,213 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+#nullable enable
+namespace EventStore.Client {
+	internal class GossipChannelSelector : IChannelSelector {
```

Comment:
looks like in this implementation of IChannelSelector SelectChannel and Rediscover need to be called in lock-step (im not certain the callers necessarily obey this)

i wonder if the interface could actually be something like 

```
	internal interface IChannelSelector : IAsyncDisposable {
		// does discovery, selects a node according to preference, gets caps
		Task<ChannelInfo> SelectChannel(CancellationToken cancellationToken);
		// creates a channel for the specified endpoint (i.e. new leader), gets caps
		Task<ChannelInfo> GetChannel(EndPoint endPoint);
	}
```